astropy/wcs/utils.py

```diff
@@ -635,193 +585,225 @@ def pixel_to_skycoord(xp, yp, wcs, origin=0, mode='all', cls=None):
-        matrix2 = matrix2[mapping]
-
-    elif classes1 != classes2:
-
-        raise ValueError("World coordinate order doesn't match and automatic matching is ambiguous")
-
-    matrix = np.matmul(matrix2.T, matrix1)
+    from ..modeling.models import SIP, InverseSIP   # here to avoid circular import
 
-    return matrix
+    # unpack params
+    crpix = params[0:2]
+    cdx = params[2:6].reshape((2, 2))
+    a_params = params[6:6+len(coeff_names)]
+    b_params = params[6+len(coeff_names):]
 
+    # assign to wcs, used for transfomations in this function
+    w_obj.wcs.cd = cdx
+    w_obj.wcs.crpix = crpix
 
-def _split_matrix(matrix):
-    """
-    Given an axis correlation matrix from a WCS object, return information about
-    the individual WCS that can be split out.
+    a_coeff, b_coeff = {}, {}
+    for i in range(len(coeff_names)):
+        a_coeff['A_' + coeff_names[i]] = a_params[i]
+        b_coeff['B_' + coeff_names[i]] = b_params[i]
 
-    The output is a list of tuples, where each tuple contains a list of
-    pixel dimensions and a list of world dimensions that can be extracted to
-    form a new WCS. For example, in the case of a spectral cube with the first
-    two world coordinates being the celestial coordinates and the third
-    coordinate being an uncorrelated spectral axis, the matrix would look like::
+    sip = SIP(crpix=crpix, a_order=order, b_order=order,
+              a_coeff=a_coeff, b_coeff=b_coeff)
+    fuv, guv = sip(u, v)
 
-        array([[ True,  True, False],
-               [ True,  True, False],
-               [False, False,  True]])
+    xo, yo = np.dot(cdx, np.array([u+fuv-crpix[0], v+guv-crpix[1]]))
 
-    and this function will return ``[([0, 1], [0, 1]), ([2], [2])]``.
-    """
+    # use all pix2world in case `projection` contains distortion table
+    x, y = w_obj.all_world2pix(lon, lat, 1) 
+    x, y = np.dot(w_obj.wcs.cd, (x-w_obj.wcs.crpix[0], y-w_obj.wcs.crpix[1]))
 
-    pixel_used = []
+    resids = np.concatenate((x-xo, y-yo))
 
-    split_info = []
+    return resids
 
-    for ipix in range(matrix.shape[1]):
-        if ipix in pixel_used:
-            continue
-        pixel_include = np.zeros(matrix.shape[1], dtype=bool)
-        pixel_include[ipix] = True
-        n_pix_prev, n_pix = 0, 1
-        while n_pix > n_pix_prev:
-            world_include = matrix[:, pixel_include].any(axis=1)
-            pixel_include = matrix[world_include, :].any(axis=0)
-            n_pix_prev, n_pix = n_pix, np.sum(pixel_include)
-        pixel_indices = list(np.nonzero(pixel_include)[0])
-        world_indices = list(np.nonzero(world_include)[0])
-        pixel_used.extend(pixel_indices)
-        split_info.append((pixel_indices, world_indices))
 
-    return split_info
+def fit_wcs_from_points(xy, world_coords, proj_point='center', 
+                        projection='TAN', sip_distortion=True, degree=4):
 
+    """ 
+    Given two matching sets of coordinates on detector and sky,
```

Comment:
The first line of the docstring should give a one sentence summary, then empty line, then full description.